rfcs/2020-09-03-3226-api-topology.md

```diff
@@ -0,0 +1,141 @@
+# RFC #3226 - 2020-09-03 - API Topology
+
+This RFC discusses configuration, implementation and potential schema for the API to co-exist with, and observe, existing Vector topology.
+
+## Scope
+
+* Implementation of the API server alongside existing topology.
+* Configuration in `vector.toml` to set-up / enable the API.
+* GraphQL types to observe existing topology (pending #3648)
+
+## Motivation
+
+I implemented the first draft of a GraphQL API server in #3514, and made changes to the toplogy mod to accommodate it. @lukesteensen [commented](https://github.com/timberio/vector/pull/3514#discussion_r479517596) with a preference to keep the API out of the existing topology mod.
+
+This raises several design questions:
+
+* Where should the API live?
+* If the API is outside of topology, what are our expectations with users making API configuration changes in `vector.toml` and handling SIGHUP/reloading? (The initial draft in #3514 )
+* Which parts of the API are configurable?
+* What are our expectations for enabling/disabling the config?
+* Should the API be treated as a sink - or exist as a layer atop topology?
+
+Additionally, there are questions about observing topology:
+
+* Should the API consume `internal_metrics` as a source - or subscribe directly to the controller, to override source configuration?
+* More generally - how will the API observe sources/sinks/transforms?
```

Comment:
The two sources will be the internal metrics themselves and probably a shared view of the current configuration. The configuration sharing will likely be as simple as a [tokio `watch` channel](https://docs.rs/tokio/0.2.22/tokio/sync/watch/index.html) that is written to during config reloading and watched by the API server.

We may eventually want to expose more direct information from specific components that do not fit a metrics data model (e.g. the status of open connections or files we're tailing), but that can wait until we've solidified the basics.

Reply:
👍 Thanks, that helps clarify the approach.